File: Python_based/gear_model/generator/trapezoidal_rack.py
"""Port of +Generator/trapezoidalRack.m."""
from __future__ import annotations

import math
import warnings

import numpy as np
from scipy.optimize import brentq

from gear_model.generator.tool_root_rounding import tool_root_rounding


class TrapezoidalRack:
    """Trapezoidal basic tooth profile for involute gearing."""

    Cs = 0.5  # Tooth thickness coefficient

    def __init__(
        self,
        modul: float = 1.0,
        profile_angle: float = 20.0,
        unit_shift: float = 0.0,
        quality: int = 25,
    ) -> None:
        self.m_n = float(modul)
        self.alpha = math.radians(profile_angle)
        self.x = float(unit_shift)
        self.X_d = self.m_n * self.x

        self.c_koef = 0.25
        self.rf_koef = 0.38

        # Tool rounding radii (CSN 01 4608)
        self.r_f0 = tool_root_rounding(self.m_n)
        self.r_a0 = self.rf_koef * self.m_n

        self.c_0 = math.ceil(4 * math.sqrt(self.m_n)) / 10.0
        self.h_a0 = self.m_n * (1 + self.c_koef)
        self.h_f0 = self.m_n + self.c_0

        self.p = self.m_n * math.pi
        self.quality = int(quality)

        # Maximum profile angle
        def alpha_eq(par: float) -> float:
            return (
                self.Cs * self.m_n * math.pi / 2
                - self.h_a0 * math.tan(par)
                - self.r_a0 * math.tan(math.pi / 4 - par / 2)
            )

        try:
            amax = brentq(alpha_eq, 1e-6, math.pi / 2 - 1e-6)
        except ValueError:
            amax = math.radians(20)
        self.alpha_max = math.floor(math.degrees(amax) * 1e4) * 1e-4

        if profile_angle > self.alpha_max:
            warnings.warn(
                f"Profile angle exceeds maximum ({self.alpha_max:.4f} deg)."
            )

        self.Cs_lower = (
            self.h_a0 * math.tan(self.alpha)
            + self.r_a0 * math.tan(math.pi / 4 - self.alpha / 2)
        ) / (self.m_n * math.pi)
        self.Cs_upper = 1 - 2 / (self.m_n * math.pi) * (
            self.h_f0 * math.tan(self.alpha)
            + self.r_f0 * math.tan(math.pi / 4 - self.alpha / 2)
        )

        self._build_profile()

    def _build_profile(self) -> None:
        # Significant profile points (2x4)
        points = np.zeros((2, 4))
        points[0, 3] = self.Cs * self.m_n * math.pi / 2
        points[0, 0] = (
            points[0, 3]
            - self.h_a0 * math.tan(self.alpha)
            - self.r_a0 * math.tan(math.pi / 4 - self.alpha / 2)
        )
        points[0, 1] = points[0, 0]
        points[0, 2] = points[0, 0] + self.r_a0 * math.cos(self.alpha)

        points[1, 0] = -self.h_a0 + self.r_a0
        points[1, 1] = -self.h_a0
        points[1, 2] = points[1, 0] - self.r_a0 * math.sin(self.alpha)
        points[1, 3] = 0.0

        # Root fillet centre
        O_f = (
            points[0, 3]
            + self.h_f0 * math.tan(self.alpha)
            + self.r_f0 * math.tan(math.pi / 4 - self.alpha / 2),
            self.h_f0 - self.r_f0,
        )

        # linspace includes endpoints (MATLAB and numpy behave identically)
        t = np.linspace(0.0, math.pi / 2 - self.alpha, int(self.quality))

        head_x = self.r_a0 * np.sin(t) + points[0, 0]
        head_y = -self.r_a0 * np.cos(t) + points[1, 0]
        root_x = -self.r_f0 * np.cos(t + self.alpha) + O_f[0]
        root_y = self.r_f0 * np.sin(t + self.alpha) + O_f[1]

        profile_x = np.concatenate(
            ([0.0], head_x, root_x, [self.m_n * math.pi / 2])
        )
        profile_y = np.concatenate(
            ([-self.h_a0], head_y, root_y, [self.h_f0])
        )
        profile = np.vstack([profile_x, profile_y])

        # Apply profile shift
        points[1, :] += self.X_d
        profile[1, :] += self.X_d

        # Mirror to full tooth
        mirror = np.vstack([-profile[0, 1:], profile[1, 1:]])[:, ::-1]
        self.tooth = np.hstack([mirror, profile[:, 1:]])

        mirror_pts = np.vstack([-points[0, :], points[1, :]])[:, ::-1]
        self.points = np.hstack([mirror_pts, points])
        self.profile = profile

File: MATLAB_based/+Generator/trapezoidalRack.m
% Copyright (c) 2026 Richard Timko
classdef trapezoidalRack
    % trapezoidalRack — Generates the trapezoidal basic tooth profile
    % for involute gearing (rack-cutter reference profile).

    properties
        m_n   (1,1) double  % Normal module [mm]
        alpha (1,1) double  % Profile angle [rad]
        x     (1,1) double  % Profile shift coefficient [-]
        Cs    (1,1) double {mustBeNumeric, mustBePositive} = 0.5 % Tooth thickness coefficient [-]
    end

    properties (SetAccess = private)
        r_a0  (1,1) double  % Tool addendum rounding radius [mm]
        r_f0  (1,1) double  % Tool dedendum rounding radius [mm]
        c_0   (1,1) double  % Tool tip clearance [mm]
        h_a0  (1,1) double  % Tool addendum height [mm]
        h_f0  (1,1) double  % Tool dedendum height [mm]
        X_d   (1,1) double  % Radial tool shift [mm]

        p     (1,1) double  % Pitch [mm]

        c_koef  (1,1) double {mustBeNumeric, mustBePositive} = 0.25
        rf_koef (1,1) double {mustBeNumeric, mustBePositive} = 0.38

        Cs_lower  (1,1) double
        Cs_upper  (1,1) double
        alpha_max (1,1) double
    end

    properties (SetAccess = private, Hidden)
        quality (1,1) uint8
        points  double
        profile double
        tooth   double
    end

    methods
        function obj = trapezoidalRack(modul, profile_angle, unit_shift, quality)
            % Constructor — create a trapezoidalRack object.
            arguments
                modul         (1,1) double {mustBeNumeric, mustBePositive} = 1
                profile_angle (1,1) double {mustBeNumeric, mustBePositive} = 20
                unit_shift    (1,1) double {mustBeNumeric}                 = 0
                quality       (1,1) uint8  {mustBeInteger}                 = 25
            end
            obj.quality = quality;

            obj.m_n   = modul;
            obj.alpha = deg2rad(profile_angle);
            obj.x     = unit_shift;
            obj.X_d   = obj.m_n * obj.x;

            obj = computeToolRounding(obj);
            obj.c_0  = ceil(4 * sqrt(obj.m_n)) / 10;
            obj.h_a0 = obj.m_n * (1 + obj.c_koef);
            obj.h_f0 = obj.m_n + obj.c_0;

            obj.p = obj.m_n * pi;

            % Maximum profile angle from the rack geometry
            obj.alpha_max = fzero(@(par) obj.Cs*obj.m_n*pi/2 - obj.h_a0*tan(par) - obj.r_a0*tan(pi/4-par/2), deg2rad(20));
            obj.alpha_max = floor(rad2deg(obj.alpha_max)*1e4) * 1e-4;

            if profile_angle > obj.alpha_max
                warning('The entered profile angle exceeds the maximum (%.4f°).', obj.alpha_max);
            end

            obj.Cs_lower = (obj.h_a0*tan(obj.alpha) + obj.r_a0*tan(pi/4-obj.alpha/2)) / (obj.m_n*pi);
            obj.Cs_upper = 1 - 2/(obj.m_n*pi) * (obj.h_f0*tan(obj.alpha) + obj.r_f0*tan(pi/4-obj.alpha/2));

            obj = profileFuncion(obj);
        end

        function obj = profileFuncion(obj)
            % Generate the trapezoidal rack tooth profile.

            % Significant profile points
            obj.points = zeros(2, 4);
            obj.points(1,4) = obj.Cs * obj.m_n * pi / 2;
            obj.points(1,1) = obj.points(1,4) - obj.h_a0*tan(obj.alpha) - obj.r_a0*tan(pi/4 - obj.alpha/2);
            obj.points(1,2) = obj.points(1,1);
            obj.points(1,3) = obj.points(1,1) + obj.r_a0*cos(obj.alpha);

            obj.points(2,1:2) = [-obj.h_a0 + obj.r_a0,  -obj.h_a0];
            obj.points(2,3)   = obj.points(2,1) - obj.r_a0*sin(obj.alpha);
            obj.points(2,4)   = 0;

            % Root fillet centre
            O_f(1) = obj.points(1,4) + obj.h_f0*tan(obj.alpha) + obj.r_f0*tan(pi/4 - obj.alpha/2);
            O_f(2) = obj.h_f0 - obj.r_f0;

            % Parameter for fillet arcs
            t = linspace(0, pi/2 - obj.alpha, obj.quality);

            % Assemble the half-tooth profile
            obj.profile = [0  obj.r_a0*sin(t)+obj.points(1,1)  -obj.r_f0*cos(t+obj.alpha)+O_f(1)  obj.m_n*pi/2; ...
                           -obj.h_a0  -obj.r_a0*cos(t)+obj.points(2,1)  obj.r_f0*sin(t+obj.alpha)+O_f(2)  obj.h_f0];

            % Apply profile shift
            obj.points(2,:) = obj.points(2,:) + obj.X_d;
            obj.profile(2,:) = obj.profile(2,:) + obj.X_d;

            % Mirror to full tooth
            obj.tooth = [fliplr([-obj.profile(1,2:end); obj.profile(2,2:end)])  obj.profile(:,2:end)];
            obj.points = [fliplr([-obj.points(1,:); obj.points(2,:)])  obj.points];
        end

        function obj = computeToolRounding(obj)
            % Determine hob tool rounding radii per CSN 01 4608
            obj.r_f0 = Generator.toolRootRounding(obj.m_n);
            obj.r_a0 = obj.rf_koef * obj.m_n;
        end
    end
end
